File: Documentation/5.4/Samples/python/ClientApi/Operations/Server/Compact.py
from ravendb import GetIndexNamesOperation, GetDatabaseRecordOperation
from ravendb.documents.operations.compact import CompactDatabaseOperation
from ravendb.documents.operations.definitions import OperationIdResult
from ravendb.primitives.constants import int_max
from ravendb.serverwide.misc import CompactSettings
from ravendb.serverwide.operations.common import ServerOperation

from examples_base import ExampleBase


class Foo:
    class CompactDatabaseOperation(ServerOperation[OperationIdResult]):
        def __init__(self, compact_settings: CompactSettings) -> None: ...


class Compact(ExampleBase):
    def setUp(self):
        super().setUp()

    def test_compact(self):
        with self.embedded_server.get_document_store("Compact") as store:
            # region compact_0
            # Define the compact settings
            settings = CompactSettings(
                # Database to compact
                "Northwind",
                # Set 'documents' to True to compact all documents in database
                # Indexes are not set and will not be compacted
                documents=True,
            )

            # Define the compact operation, pass the settings
            compact_op = CompactDatabaseOperation(settings)

            # Execute compaction by passing the operation to maintenance.server.send
            operation = store.maintenance.server.send_async(compact_op)

            # Wait for operation to complete, during compaction the database is offline
            operation.wait_for_completion()
            # endregion

            # region compact_1
            # Define the compact settings
            settings = CompactSettings(
                # Database to compact
                database_name="Northwind",
                # Setting 'documents' to False will compact only the specified indexes
                documents=False,
                # Specify which indexes to compact
                indexes=["Orders/Totals", "Orders/ByCompany"],
                # Optimize indexes is Lucene's feature to gain disk space and efficiency
                # Set whether to skip this optimization when compacting the indexes
                skip_optimize_indexes=False,
            )
            # Define the compact operation, pass the settings
            compact_op = CompactDatabaseOperation(settings)

            # Execute compaction by passing the operation to maintenance.server.send
            operation = store.maintenance.server.send_async(compact_op)
            # Wait for operation to complete
            operation.wait_for_completion()
            # endregion

            # region compact_2
            #  Get all indexes names in the database using the 'GetIndexNamesOperation' operation
            #  Use 'ForDatabase' if the target database is different from the default database defined on the store
            all_indexes_names = store.maintenance.for_database("Northwind").send(GetIndexNamesOperation(0, int_max))

            # Define the compact settings
            settings = CompactSettings(
                database_name="Northwind",  # Database to compact
                documents=True,  # Compact all documents
                indexes=all_indexes_names,  # All indexes will be compacted
                skip_optimize_indexes=True,  # Skip Lucene's indexes optimization
            )

            # Define the compact operation, pass the settings
            compact_op = CompactDatabaseOperation(settings)

            # Execute compaction by passing the operation to maintenance.server.send
            operation = store.maintenance.server.send(compact_op)
            # Wait for operation to complete
            operation.wait_for_completion()
            # endregion

            # region compact_3
            # Get all member nodes in the database-group using the 'GetDatabaseRecordOperation' operation
            all_member_nodes = store.maintenance.server.send(GetDatabaseRecordOperation("Northwind")).topology.members

            # Define the compact settings as needed
            settings = CompactSettings(
                # Database to compact
                database_name="Northwind",
                # Compact all documents in database
                documents=True,
            )

            # Execute the compact operation on each member node
            for node_tag in all_member_nodes:
                # Define the compact operation, pass the settings
                compact_op = CompactDatabaseOperation(settings)

                # Execute the operation on a specific node
                # Use 'for_node' to specify the node to operate on
                # todo: https://issues.hibernatingrhinos.com/issue/RDBC-847/Implement-documentStore.Maintenance.Server.ForNode
                operation = store.maintenance.server.for_node(node_tag).send_async(compact_op)

                # Wait for operation to complete
                operation.wait_for_completion()

            # endregion

File: Documentation/5.4/Samples/php/ClientApi/Operations/Server/Compact.php
<?php

namespace RavenDB\Samples\ClientApi\Operations\Server;

use RavenDB\Constants\PhpClient;
use RavenDB\Documents\DocumentStore;
use RavenDB\Documents\Operations\Indexes\GetIndexNamesOperation;
use RavenDB\Documents\Operations\Operation;
use RavenDB\Documents\Operations\OperationIdResult;
use RavenDB\Documents\Operations\CompactDatabaseOperation;
use RavenDB\ServerWide\CompactSettings;
use RavenDB\ServerWide\DatabaseRecordWithEtag;
use RavenDB\ServerWide\GetDatabaseRecordOperation;
use RavenDB\Type\StringArrayResult;

class Compact
{
    public function samples(): void
    {
        $documentStore = new DocumentStore();
        try {
            # region compact_0

            // Define the compact settings
            $settings = new CompactSettings();
            $settings->setDatabaseName("Northwind");
            // Set 'Documents' to true to compact all documents in database
            // Indexes are not set and will not be compacted
            $settings->setDocuments(true);


            // Define the compact operation, pass the settings
            /** @var OperationIdResult  $compactOp */
            $compactOp = new CompactDatabaseOperation($settings);

            // Execute compaction by passing the operation to Maintenance.Server.Send
            /** @var Operation $operation */
            $operation = $documentStore->maintenance()->server()->send($compactOp);

            // Wait for operation to complete, during compaction the database is offline
            $operation->waitForCompletion();
            # endregion
        } finally {
            $documentStore->close();
        }

        $documentStore = new DocumentStore();
        try {
            # region compact_1
            // Define the compact settings
            $settings = new CompactSettings();

            // Database to compact
            $settings->setDatabaseName("Northwind");

            // Setting 'Documents' to false will compact only the specified indexes
            $settings->setDocuments(false);

            // Specify which indexes to compact
            $settings->setIndexes([ "Orders/Totals", "Orders/ByCompany" ]);

            // Optimize indexes is Lucene's feature to gain disk space and efficiency
            // Set whether to skip this optimization when compacting the indexes
            $settings->setSkipOptimizeIndexes(false);


            // Define the compact operation, pass the settings
            /** @var OperationIdResult $compactOp */
            $compactOp = new CompactDatabaseOperation($settings);

            // Execute compaction by passing the operation to Maintenance.Server.Send
            /** @var Operation $operation */
            $operation = $documentStore->maintenance()->server()->send($compactOp);
            // Wait for operation to complete
            $operation->waitForCompletion();
            # endregion
            } finally {
                $documentStore->close();
            }


        $documentStore = new DocumentStore();
        try {
            # region compact_2
            // Get all indexes names in the database using the 'GetIndexNamesOperation' operation
            // Use 'ForDatabase' if the target database is different than the default database defined on the store
            /** @var StringArrayResult $allIndexNames */
            $allIndexNames = $documentStore->maintenance()->forDatabase("Northwind")
                    ->send(new GetIndexNamesOperation(0, PhpClient::INT_MAX_VALUE));

            // Define the compact settings
            $settings = new CompactSettings();
            $settings->setDatabaseName("Northwind");    // Database to compact
            $settings->setDocuments(true);              // Compact all documents
            $settings->setIndexes($allIndexNames->getArrayCopy());      // All indexes will be compacted
            $settings->setSkipOptimizeIndexes(true);    // Skip Lucene's indexes optimization

            // Define the compact operation, pass the settings
            /** @var OperationIdResult $compactOp */
            $compactOp = new CompactDatabaseOperation($settings);

            // Execute compaction by passing the operation to Maintenance.Server.Send
            /** @var Operation $operation */
            $operation = $documentStore->maintenance()->server()->send($compactOp);

            // Wait for operation to complete
            $operation->waitForCompletion();
            # endregion
        } finally {
            $documentStore->close();
        }

        $documentStore = new DocumentStore();
        try {
            # region compact_3
            // Get all member nodes in the database-group using the 'GetDatabaseRecordOperation' operation
            /** @var DatabaseRecordWithEtag $databaseRecord */
            $databaseRecord = $documentStore->maintenance()->server()->send(new GetDatabaseRecordOperation("Northwind"));

            $allMemberNodes = $databaseRecord->getTopology()->getMembers();

            // Define the compact settings as needed
            $settings = new CompactSettings();

            $settings->setDatabaseName("Northwind");
            $settings->setDocuments(true); //Compact all documents in database

            // Execute the compact operation on each member node
            foreach ($allMemberNodes as $nodeTag) {
                // Define the compact operation, pass the settings
                /** @var OperationIdResult $compactOp */
                $compactOp = new CompactDatabaseOperation($settings);

                // Execute the operation on a specific node
                // Use `ForNode` to specify the node to operate on
                /** @var Operation $operation */
                $operation = $documentStore->maintenance()->server()->forNode($nodeTag)->send($compactOp);
                // Wait for operation to complete
                $operation->waitForCompletion();
            }
            # endregion
        } finally {
            $documentStore->close();
        }
    }
}

/*
interface IFoo
{
    # region syntax
    CompactDatabaseOperation(?CompactSettings $compactSettings)
    # endregion
}
*/
